File: hekatan-fem/test_wilson_membrane.cpp
// =============================================================================
// Test standalone C++ — Wilson 4 incompatible modes Q4 membrane
//                       + Taylor 1976 J0 patch test correction.
//
// Equivalente exacto del test Python test_wilson_membrane.py.
// Genera test_wilson_cpp_result.json para comparar con test_wilson_python_result.json.
//
// Compilar:
//   g++ -O2 -std=c++17 -I src/cpp/eigen test_wilson_membrane.cpp -o test_wilson_membrane.exe
// Correr:
//   PATH="/c/Program Files/GNU Octave/Octave-10.1.0/mingw64/bin:$PATH" ./test_wilson_membrane.exe
// =============================================================================

#include <iostream>
#include <iomanip>
#include <fstream>
#include <cmath>
#include <vector>
#include <Eigen/Dense>

static const double GP = 0.5773502691896258;  // 1/sqrt(3)
static const double GP2x2[4][2] = {
    {-GP, -GP}, {GP, -GP}, {GP, GP}, {-GP, GP}
};

static void shape_q4(double xi, double eta,
                      double N[4], double dNdxi[4], double dNdeta[4]) {
    N[0] = 0.25 * (1 - xi) * (1 - eta);
    N[1] = 0.25 * (1 + xi) * (1 - eta);
    N[2] = 0.25 * (1 + xi) * (1 + eta);
    N[3] = 0.25 * (1 - xi) * (1 + eta);
    dNdxi[0]  = -0.25 * (1 - eta); dNdxi[1]  =  0.25 * (1 - eta);
    dNdxi[2]  =  0.25 * (1 + eta); dNdxi[3]  = -0.25 * (1 + eta);
    dNdeta[0] = -0.25 * (1 - xi);  dNdeta[1] = -0.25 * (1 + xi);
    dNdeta[2] =  0.25 * (1 + xi);  dNdeta[3] =  0.25 * (1 - xi);
}

static double jacobian2D(const double x[4], const double y[4],
                          const double dNdxi[4], const double dNdeta[4],
                          double Jinv[2][2]) {
    double J00 = 0, J01 = 0, J10 = 0, J11 = 0;
    for (int i = 0; i < 4; i++) {
        J00 += dNdxi[i]  * x[i];
        J01 += dNdxi[i]  * y[i];
        J10 += dNdeta[i] * x[i];
        J11 += dNdeta[i] * y[i];
    }
    double detJ = J00 * J11 - J01 * J10;
    if (std::abs(detJ) < 1e-15) detJ = 1e-15;
    double inv = 1.0 / detJ;
    Jinv[0][0] =  J11 * inv;
    Jinv[0][1] = -J01 * inv;
    Jinv[1][0] = -J10 * inv;
    Jinv[1][1] =  J00 * inv;
    return detJ;
}

// Q4 plane stress estándar (sin Wilson) — equivalente a shellThin.cpp::getMembraneK_Thin
Eigen::MatrixXd membrane_K_plain(const double x[4], const double y[4],
                                   double E, double nu, double t) {
    double factor = E / (1.0 - nu * nu);
    Eigen::Matrix3d Dm;
    Dm << factor,       factor * nu, 0,
          factor * nu,  factor,      0,
          0,            0,           factor * (1 - nu) / 2.0;

    Eigen::MatrixXd K = Eigen::MatrixXd::Zero(8, 8);
    for (int gp = 0; gp < 4; gp++) {
        double xi = GP2x2[gp][0], eta = GP2x2[gp][1];
        double N[4], dNdxi[4], dNdeta[4];
        shape_q4(xi, eta, N, dNdxi, dNdeta);
        double Jinv[2][2];
        double detJ = jacobian2D(x, y, dNdxi, dNdeta, Jinv);

        Eigen::MatrixXd B = Eigen::MatrixXd::Zero(3, 8);
        for (int i = 0; i < 4; i++) {
            double dNdx = Jinv[0][0] * dNdxi[i] + Jinv[0][1] * dNdeta[i];
            double dNdy = Jinv[1][0] * dNdxi[i] + Jinv[1][1] * dNdeta[i];
            B(0, 2*i)     = dNdx;
            B(1, 2*i + 1) = dNdy;
            B(2, 2*i)     = dNdy;
            B(2, 2*i + 1) = dNdx;
        }
        K += (t * std::abs(detJ)) * B.transpose() * Dm * B;
    }
    return K;
}

// Wilson 1971 + Taylor 1976 — equivalente a shellQ4.cpp::getMembraneK
Eigen::MatrixXd membrane_K_wilson(const double x[4], const double y[4],
                                    double E, double nu, double t) {
    double factor = E / (1.0 - nu * nu);
    Eigen::Matrix3d Dm;
    Dm << factor,       factor * nu, 0,
          factor * nu,  factor,      0,
          0,            0,           factor * (1 - nu) / 2.0;

    // J0 evaluado en centro (ξ=η=0)
    double Jinv0[2][2];
    {
        double N0[4], dN0dxi[4], dN0deta[4];
        shape_q4(0.0, 0.0, N0, dN0dxi, dN0deta);
        jacobian2D(x, y, dN0dxi, dN0deta, Jinv0);
    }

    Eigen::MatrixXd Kuu = Eigen::MatrixXd::Zero(8, 8);
    Eigen::MatrixXd Kua = Eigen::MatrixXd::Zero(8, 4);
    Eigen::MatrixXd Kaa = Eigen::MatrixXd::Zero(4, 4);

    for (int gp = 0; gp < 4; gp++) {
        double xi = GP2x2[gp][0], eta = GP2x2[gp][1];
        double N[4], dNdxi[4], dNdeta[4];
        shape_q4(xi, eta, N, dNdxi, dNdeta);
        double Jinv[2][2];
        double detJ = jacobian2D(x, y, dNdxi, dNdeta, Jinv);

        // Bc 3×8 compatible
        Eigen::MatrixXd Bc = Eigen::MatrixXd::Zero(3, 8);
        for (int i = 0; i < 4; i++) {
            double dNdx = Jinv[0][0] * dNdxi[i] + Jinv[0][1] * dNdeta[i];
            double dNdy = Jinv[1][0] * dNdxi[i] + Jinv[1][1] * dNdeta[i];
            Bc(0, 2*i)     = dNdx;
            Bc(1, 2*i + 1) = dNdy;
            Bc(2, 2*i)     = dNdy;
            Bc(2, 2*i + 1) = dNdx;
        }

        // Bi 3×4 incompatible (N5=1-ξ², N6=1-η², derivadas con J0)
        double dN5dxi  = -2.0 * xi,  dN5deta = 0.0;
        double dN6dxi  =  0.0,       dN6deta = -2.0 * eta;
        double dN5dx = Jinv0[0][0] * dN5dxi + Jinv0[0][1] * dN5deta;
        double dN5dy = Jinv0[1][0] * dN5dxi + Jinv0[1][1] * dN5deta;
        double dN6dx = Jinv0[0][0] * dN6dxi + Jinv0[0][1] * dN6deta;
        double dN6dy = Jinv0[1][0] * dN6dxi + Jinv0[1][1] * dN6deta;

        Eigen::MatrixXd Bi = Eigen::MatrixXd::Zero(3, 4);
        Bi(0, 0) = dN5dx;
        Bi(0, 1) = dN6dx;
        Bi(1, 2) = dN5dy;
        Bi(1, 3) = dN6dy;
        Bi(2, 0) = dN5dy;
        Bi(2, 1) = dN6dy;
        Bi(2, 2) = dN5dx;
        Bi(2, 3) = dN6dx;

        double w = t * std::abs(detJ);
        Kuu += w * Bc.transpose() * Dm * Bc;
        Kua += w * Bc.transpose() * Dm * Bi;
        Kaa += w * Bi.transpose() * Dm * Bi;
    }

    Eigen::MatrixXd KaaInv;
    if (std::abs(Kaa.determinant()) > 1e-20) {
        KaaInv = Kaa.inverse();
    } else {
        std::cerr << "Warning: Kaa singular, fallback Kuu" << std::endl;
        return Kuu;
    }
    return Kuu - Kua * KaaInv * Kua.transpose();
}

int main() {
    // Mismo Q4 unit del Python: 1m × 1m
    double x[4] = {0.0, 1.0, 1.0, 0.0};
    double y[4] = {0.0, 0.0, 1.0, 1.0};
    double E = 24.85e6;
    double nu = 0.15;
    double t = 0.10;

    std::cout << std::scientific << std::setprecision(6);
    std::cout << "======================================================================\n";
    std::cout << "Test Wilson Q4 Membrane - C++ standalone (Eigen)\n";
    std::cout << "  Q4 unit 1m x 1m, E=" << E << ", nu=" << nu << ", t=" << t << "\n";
    std::cout << "======================================================================\n\n";

    Eigen::MatrixXd K_plain  = membrane_K_plain(x, y, E, nu, t);
    Eigen::MatrixXd K_wilson = membrane_K_wilson(x, y, E, nu, t);

    std::cout << "[K_plain]   norma Frobenius = " << K_plain.norm()  << "\n";
    std::cout << "[K_wilson]  norma Frobenius = " << K_wilson.norm() << "\n";
    std::cout << "\nK_plain[0,0]   = " << K_plain(0, 0)  << "\n";
    std::cout << "K_wilson[0,0]  = " << K_wilson(0, 0) << "\n";
    std::cout << "Ratio (Wilson/Plain) [0,0] = " << K_wilson(0, 0) / K_plain(0, 0) << "\n";

    // Cantilever test (mismo del Python)
    std::cout << "\n======================================================================\n";
    std::cout << "Test cantilever (bending in-plane): empotrar nodos 0,1; F=1 horizontal en 2,3\n";
    std::cout << "======================================================================\n";

    auto solve_cantilever = [](const Eigen::MatrixXd &K, const char *label) {
        // Free DOFs: u2=4, v2=5, u3=6, v3=7
        Eigen::MatrixXd Kr(4, 4);
        int idx[4] = {4, 5, 6, 7};
        for (int i = 0; i < 4; i++)
            for (int j = 0; j < 4; j++)
                Kr(i, j) = K(idx[i], idx[j]);
        Eigen::VectorXd F(4);
        F << 0.5, 0.0, 0.5, 0.0;
        Eigen::VectorXd ur = Kr.fullPivLu().solve(F);
        double u2 = ur(0);
        std::cout << "  [" << label << "]  u_top (u2 ~= u3) = " << u2 << " m\n";
        return u2;
    };

    double u_plain  = solve_cantilever(K_plain,  "Plain Q4 ");
    double u_wilson = solve_cantilever(K_wilson, "Wilson Q4");
    std::cout << "\n  Ratio Wilson/Plain = " << u_wilson / u_plain << "\n";

    // Save to JSON
    std::ofstream out("test_wilson_cpp_result.json");
    out << std::scientific << std::setprecision(15);
    out << "{\n";
    out << "  \"input\": {\"x\": [" << x[0] << "," << x[1] << "," << x[2] << "," << x[3]
        << "], \"y\": [" << y[0] << "," << y[1] << "," << y[2] << "," << y[3]
        << "], \"E\": " << E << ", \"nu\": " << nu << ", \"t\": " << t << "},\n";
    auto dumpMat = [&](const Eigen::MatrixXd &M, const char *name, bool last) {
        out << "  \"" << name << "\": [";
        for (int i = 0; i < M.rows(); i++) {
            out << "[";
            for (int j = 0; j < M.cols(); j++) {
                out << M(i, j);
                if (j < M.cols() - 1) out << ",";
            }
            out << "]";
            if (i < M.rows() - 1) out << ",";
        }
        out << "]";
        if (!last) out << ",";
        out << "\n";
    };
    dumpMat(K_plain,  "K_plain",  false);
    dumpMat(K_wilson, "K_wilson", false);
    out << "  \"u_plain_top\":  " << u_plain  << ",\n";
    out << "  \"u_wilson_top\": " << u_wilson << "\n";
    out << "}\n";
    out.close();
    std::cout << "\nResultados guardados en test_wilson_cpp_result.json\n";
    return 0;
}

File: hekatan-fem/test_wilson_membrane.py
"""
Test standalone Python — Wilson 4 incompatible modes Q4 membrane
                          + Taylor 1976 J0 patch test correction.

Referencia académica:
- Wilson, Taylor, Doherty, Ghaboussi (1973), "Incompatible Displacement Models"
- Taylor, Beresford, Wilson (1976), "A non-conforming element for stress analysis"
- Cook, Malkus, Plesha, Witt (2002), §6.6 "Incompatible Elements"
- CSI Analysis Reference Manual §10.1.1 (formulación de SAP2000/ETABS)

Esta implementación es la referencia "ground truth" para validar el C++ equivalente.

Funciones de forma extendidas (4 modos α internos):
  u_x(ξ,η) = Σ N_i(ξ,η)·u_xi + α₁·(1−ξ²) + α₂·(1−η²)
  u_y(ξ,η) = Σ N_i(ξ,η)·u_yi + α₃·(1−ξ²) + α₄·(1−η²)

Taylor 1976 patch test: las derivadas de N₅=(1−ξ²) y N₆=(1−η²) se evalúan
con el Jacobiano J₀ EVALUADO EN EL CENTRO (ξ=η=0), no con J(ξ,η). Esto
fuerza ∫B_I dV = 0 → pasa patch test en mallas distorsionadas.

Condensación estática: K_uu - K_uα · K_αα⁻¹ · K_αu → 8×8 final.
"""

import numpy as np


# 2×2 Gauss points
GP = 1.0 / np.sqrt(3.0)
GP2x2 = [(-GP, -GP), (GP, -GP), (GP, GP), (-GP, GP)]


def shape_q4(xi, eta):
    """N[4], dN/dxi[4], dN/deta[4] para Q4 bilineal CCW."""
    N = np.array([
        0.25 * (1 - xi) * (1 - eta),
        0.25 * (1 + xi) * (1 - eta),
        0.25 * (1 + xi) * (1 + eta),
        0.25 * (1 - xi) * (1 + eta),
    ])
    dNdxi = np.array([
        -0.25 * (1 - eta),
         0.25 * (1 - eta),
         0.25 * (1 + eta),
        -0.25 * (1 + eta),
    ])
    dNdeta = np.array([
        -0.25 * (1 - xi),
        -0.25 * (1 + xi),
         0.25 * (1 + xi),
         0.25 * (1 - xi),
    ])
    return N, dNdxi, dNdeta


def jacobian2D(x, y, dNdxi, dNdeta):
    """Devuelve detJ, Jinv 2×2."""
    J = np.zeros((2, 2))
    for i in range(4):
        J[0, 0] += dNdxi[i] * x[i]
        J[0, 1] += dNdxi[i] * y[i]
        J[1, 0] += dNdeta[i] * x[i]
        J[1, 1] += dNdeta[i] * y[i]
    detJ = J[0, 0] * J[1, 1] - J[0, 1] * J[1, 0]
    if abs(detJ) < 1e-15:
        detJ = 1e-15
    Jinv = np.array([
        [ J[1, 1] / detJ, -J[0, 1] / detJ],
        [-J[1, 0] / detJ,  J[0, 0] / detJ],
    ])
    return detJ, Jinv


def membrane_K_plain(x, y, E, nu, t):
    """
    Q4 plane stress estándar SIN Wilson modes — la formulación actual de
    shellThin.cpp::getMembraneK_Thin. Para comparación.
    K 8×8 sobre DOFs [u0,v0, u1,v1, u2,v2, u3,v3].
    """
    factor = E / (1.0 - nu * nu)
    Dm = factor * np.array([
        [1,  nu, 0],
        [nu, 1,  0],
        [0,  0,  (1 - nu) / 2],
    ])

    K = np.zeros((8, 8))
    for xi, eta in GP2x2:
        N, dNdxi, dNdeta = shape_q4(xi, eta)
        detJ, Jinv = jacobian2D(x, y, dNdxi, dNdeta)

        B = np.zeros((3, 8))
        for i in range(4):
            dNdx = Jinv[0, 0] * dNdxi[i] + Jinv[0, 1] * dNdeta[i]
            dNdy = Jinv[1, 0] * dNdxi[i] + Jinv[1, 1] * dNdeta[i]
            B[0, 2*i]     = dNdx
            B[1, 2*i + 1] = dNdy
            B[2, 2*i]     = dNdy
            B[2, 2*i + 1] = dNdx

        w = t * abs(detJ)
        K += w * B.T @ Dm @ B
    return K


def membrane_K_wilson(x, y, E, nu, t):
    """
    Wilson 1971 + Taylor 1976 — Q4 plane stress con 4 incompatible modes
    (α₁, α₂, α₃, α₄) condensados estáticamente. K 8×8.

    Implementación exacta de hekatan-fem/src/cpp/utils/shellQ4.cpp::getMembraneK.
    """
    factor = E / (1.0 - nu * nu)
    Dm = factor * np.array([
        [1,  nu, 0],
        [nu, 1,  0],
        [0,  0,  (1 - nu) / 2],
    ])

    # J0 en el centro (ξ=η=0) — Taylor 1976 patch test correction
    _, _, _ = shape_q4(0.0, 0.0)
    _, _, dN0deta = shape_q4(0.0, 0.0)
    _, dN0dxi, _ = shape_q4(0.0, 0.0)
    # (en realidad shape_q4(0,0) ya da dN0dxi y dN0deta del centro)
    _, dN0dxi, dN0deta = shape_q4(0.0, 0.0)
    _, Jinv0 = jacobian2D(x, y, dN0dxi, dN0deta)

    Kuu = np.zeros((8, 8))
    Kua = np.zeros((8, 4))
    Kaa = np.zeros((4, 4))

    for xi, eta in GP2x2:
        N, dNdxi, dNdeta = shape_q4(xi, eta)
        detJ, Jinv = jacobian2D(x, y, dNdxi, dNdeta)

        # Bc 3×8 compatible (standard Q4)
        Bc = np.zeros((3, 8))
        for i in range(4):
            dNdx = Jinv[0, 0] * dNdxi[i] + Jinv[0, 1] * dNdeta[i]
            dNdy = Jinv[1, 0] * dNdxi[i] + Jinv[1, 1] * dNdeta[i]
            Bc[0, 2*i]     = dNdx
            Bc[1, 2*i + 1] = dNdy
            Bc[2, 2*i]     = dNdy
            Bc[2, 2*i + 1] = dNdx

        # Bi 3×4 incompatible modes (N5=1-ξ², N6=1-η², derivadas con J0)
        # N5 natural: dN5/dxi = -2ξ, dN5/deta = 0
        # N6 natural: dN6/dxi = 0, dN6/deta = -2η
        dN5dxi  = -2.0 * xi
        dN5deta =  0.0
        dN6dxi  =  0.0
        dN6deta = -2.0 * eta
        dN5dx = Jinv0[0, 0] * dN5dxi + Jinv0[0, 1] * dN5deta
        dN5dy = Jinv0[1, 0] * dN5dxi + Jinv0[1, 1] * dN5deta
        dN6dx = Jinv0[0, 0] * dN6dxi + Jinv0[0, 1] * dN6deta
        dN6dy = Jinv0[1, 0] * dN6dxi + Jinv0[1, 1] * dN6deta

        # α-ordering: [α1=u·N5, α2=u·N6, α3=v·N5, α4=v·N6]
        Bi = np.zeros((3, 4))
        Bi[0, 0] = dN5dx
        Bi[0, 1] = dN6dx
        Bi[1, 2] = dN5dy
        Bi[1, 3] = dN6dy
        Bi[2, 0] = dN5dy
        Bi[2, 1] = dN6dy
        Bi[2, 2] = dN5dx
        Bi[2, 3] = dN6dx

        w = t * abs(detJ)
        Kuu += w * Bc.T @ Dm @ Bc
        Kua += w * Bc.T @ Dm @ Bi
        Kaa += w * Bi.T @ Dm @ Bi

    # Condensación estática: K_cond = K_uu - K_uα · K_αα⁻¹ · K_αu
    try:
        Kaa_inv = np.linalg.inv(Kaa)
    except np.linalg.LinAlgError:
        print("Warning: Kaa singular, devolviendo Kuu sin condensación")
        return Kuu
    K = Kuu - Kua @ Kaa_inv @ Kua.T
    return K


def main():
    # Q4 unit cuadrado 1×1 (caso simple)
    x = np.array([0.0, 1.0, 1.0, 0.0])
    y = np.array([0.0, 0.0, 1.0, 1.0])
    E = 24.85e6    # kN/m² (concreto 4000 psi)
    nu = 0.15
    t = 0.10       # m

    print("=" * 70)
    print("Test Wilson Q4 Membrane — Python standalone")
    print(f"  Q4 unit 1m×1m, E={E:.3e}, nu={nu}, t={t}")
    print("=" * 70)

    K_plain = membrane_K_plain(x, y, E, nu, t)
    K_wilson = membrane_K_wilson(x, y, E, nu, t)

    print("\n[K_plain]   norma Frobenius =", np.linalg.norm(K_plain))
    print("[K_wilson]  norma Frobenius =", np.linalg.norm(K_wilson))
    print(f"\nK_plain[0,0]   = {K_plain[0, 0]:.6e}")
    print(f"K_wilson[0,0]  = {K_wilson[0, 0]:.6e}")
    print(f"Ratio (Wilson/Plain) [0,0] = {K_wilson[0, 0] / K_plain[0, 0]:.6f}")
    print(f"\nK_plain[0,2]   = {K_plain[0, 2]:.6e}  (acoplamiento u0-u1)")
    print(f"K_wilson[0,2]  = {K_wilson[0, 2]:.6e}")

    # ── Test cantilever pure bending: caso clásico Wilson ──
    # Aplica fuerza horizontal F en nodos 2,3 (top), restringe nodos 0,1 (bottom)
    # Para Q4 plain → underestima deflexión (locking)
    # Para Wilson → resultado mejor (más cerca del analítico Bernoulli)
    print("\n" + "=" * 70)
    print("Test cantilever puro (bending in-plane):")
    print("  Empotrar nodos 0,1 (y=0). Aplicar F=1 kN horizontal en nodos 2,3 (y=1).")
    print("=" * 70)

    def solve_cantilever(K, label):
        # Free DOFs: u2(idx 4), v2(idx 5), u3(idx 6), v3(idx 7)
        free = [4, 5, 6, 7]
        F = np.zeros(4)
        F[0] = 0.5  # u2
        F[2] = 0.5  # u3 (total F=1.0 horizontal arriba)
        Kr = K[np.ix_(free, free)]
        ur = np.linalg.solve(Kr, F)
        u2 = ur[0]  # u2 desplazamiento horizontal
        print(f"  [{label}]  u_top (u2 ~= u3) = {u2:.6e} m")
        return u2

    u_plain  = solve_cantilever(K_plain,  "Plain Q4")
    u_wilson = solve_cantilever(K_wilson, "Wilson Q4")
    print(f"\n  Ratio Wilson/Plain = {u_wilson / u_plain:.4f}")
    print(f"  -> si > 1, Wilson da MAS deflexion = MENOS locking (correcto)")

    # ── Save K matrices to JSON for C++ comparison ──
    import json
    out = {
        "input": {"x": x.tolist(), "y": y.tolist(), "E": E, "nu": nu, "t": t},
        "K_plain":  K_plain.tolist(),
        "K_wilson": K_wilson.tolist(),
        "u_plain_top":  float(u_plain),
        "u_wilson_top": float(u_wilson),
    }
    with open("test_wilson_python_result.json", "w") as f:
        json.dump(out, f, indent=2)
    print("\nResultados guardados en test_wilson_python_result.json")


if __name__ == "__main__":
    main()
